guierrorcheck refactor - round 1

NicholasBellucci opened the pull request:
Refactor to support Gui Error Check.

gilmoskowitz commented on guiclient/glSeries.cpp:
```diff
@@ -199,13 +200,6 @@ enum SetResponse glSeries::set(const ParameterList &pParams)
-    return;
-  }
 
   ParameterList params;
```
This is a fine place to use GuiErrorCheck. _Removing_ the check causes a problem - it's here to ensure that `newdlg.set(params);` a few lines later gets its required input parameter.

gilmoskowitz commented on guiclient/glSeries.cpp:
```diff
@@ -229,13 +223,6 @@ void glSeries::sNew()
-                              tr("<p>You must enter a Distribution Date for this Series.") );
-    _date->setFocus();
-    return;
-  }
```
Same here - this is an important check for valid parameter passing to a new dialog.

gilmoskowitz commented on guiclient/glTransaction.cpp:
```diff
@@ -98,37 +99,21 @@ enum SetResponse glTransaction::set(const ParameterList &pParams)
+          << GuiErrorCheck(! _credit->isValid(), _credit,
+                           tr("You must select a Credit Account for this G/L Transaction before you may Post it."))
+          << GuiErrorCheck(!_metrics->boolean("IgnoreCompany") && _credit->companyId() != _debit->companyId(), _credit,
+                           tr("The Accounts must belong to the same Company to Post this transaciton."))
```
There's a typo in the original - `transaciton` - that is being propagated here.

gilmoskowitz commented on guiclient/glTransaction.cpp:
```diff
@@ -98,37 +99,21 @@ enum SetResponse glTransaction::set(const ParameterList &pParams)
+          << GuiErrorCheck(!_metrics->boolean("IgnoreCompany") && _credit->companyId() != _debit->companyId(), _credit,
+                           tr("The Accounts must belong to the same Company to Post this transaciton."))
+          << GuiErrorCheck(_metrics->boolean("MandatoryGLEntryNotes") && _notes->toPlainText().trimmed().isEmpty(), _notes,
+                           tr("You must enter some Notes to describe this transaction."))
```
👍 You found the intermediate style!

gilmoskowitz commented on guiclient/incident.cpp:
```diff
@@ -353,35 +354,18 @@ bool incident::save(bool partial)
+          << GuiErrorCheck(_description->text().trimmed().isEmpty(), _description,
+                           tr("You must specify a description for this incident report."))
+          << GuiErrorCheck(_status->currentIndex() == 3 && _assignedTo->username().isEmpty(), _assignedTo,
+                           tr("You must specify an assignee when the status is assigned."))
```
Nice catch on the wrong focus widget.

gilmoskowitz commented on guiclient/invoice.cpp:
```diff
@@ -540,30 +541,20 @@ void invoice::sSave()
-    },
     // TODO: add more error checks here?
     { true, "", NULL }
   };
```
The whole `struct` definition can be removed.

gilmoskowitz commented on guiclient/miscCheck.cpp:
```diff
@@ -103,49 +104,21 @@ void miscCheck::sSave()
+                           tr("You must enter a date for this payment."))
+          << GuiErrorCheck(_amount->isZero(), _amount,
+                           tr("You must enter an amount for this payment."))
+          << GuiErrorCheck((_applytocm->isChecked()) && (_cmCluster->id() == -1), _expcat,
```
Hmmm... Try using `_cmCluster` as the focus widget. It would be a better choice here; I don't know why the old code had `_expcat`.

gilmoskowitz commented on guiclient/miscCheck.cpp:
```diff
@@ -103,49 +104,21 @@ void miscCheck::sSave()
-    QMessageBox::warning( this, tr("Invalid Amount"),
-                            tr("<p>You must enter an amount less than or equal to the  "
-	   		                   "credit memo selected.") );
-    _amount->setLocalValue(_aropenamt);
```
There's no good way to carry through this functionality to the new style. Thoughts?

NicholasBellucci replied:
I'm confused with this statement. I currently have it using the new functionality. Do we not want this?

gilmoskowitz replied:
The code transformation was done correctly but with a loss of functionality - the new version does not change the value of the `_amount` widget to `_aropenamt`.

That raises two questions: 
1. Is changing the value of `_amount` a feature worth preserving?
2. Was removing it a conscious decision?

To answer the first question, "No, this is not worth preserving." Rationale: the client as a rule does not change erroneous values to other potentially erroneous values during validation. Doing so loses information that might be important to the user; in this case it might be a simple typo like an extra trailing 0.

It is, however, a change from previous behavior that `git blame` says goes back to Nov 2007. A behavior change like this should be thought through carefully.

NicholasBellucci replied:
I did not notice the setLocalValue. I do believe I could get this functionality to work with the new setup if we want to preserve the feature.

gilmoskowitz replied:
Let's keep the new code here as is. I think the "feature" does more harm than good.

gilmoskowitz commented on guiclient/reconcileBankaccount.cpp:
```diff
@@ -185,49 +186,26 @@ bool reconcileBankaccount::sSave(bool closeWhenDone)
-                   "reconciled dates in it. Please choose a different "
-                   "date range.") );
-    _startDate->setFocus();
+  if(!_datesAreOK){
```
Is there a reason to keep the `if`?

NicholasBellucci replied:
I kept the if statement because in the previous setup if !_datesAreOK was a much longer function and at the end set datesAreOK to false so in order to keep that consistent I kept the if statement.

gilmoskowitz replied:
The resulting code is this:
```
  if(!_datesAreOK){
    _datesAreOK = false;
  }
```
which does nothing useful. The old code wasn't any better but the superfluous line was hidden by the bulk of the old code (going back to 2008).

gilmoskowitz commented on guiclient/standardJournal.cpp:
```diff
@@ -137,14 +132,16 @@ void standardJournal::sSave()
+  errors<< GuiErrorCheck(standardSave.first(), _name,
+                         tr("The Name you have entered for this Standard Journal already exists. "
+                           "Please enter in a different Name for this Standard Journal."))
+  ;
```
Nice catch with `standardSave.first()`.

NicholasBellucci commented:
Gil, 

I have added thoughts and comments in line with the previous email. These comments are in green. Let me know if you can’t see them.

-Nicholas Bellucci

> On Jul 31, 2017, at 6:03 PM, Gil Moskowitz <notifications@github.com> wrote:
> 
> @gilmoskowitz requested changes on this pull request.
> 
> Very nice start. Just a few minor tweaks and this will be ready to merge.
> 
> In guiclient/glSeries.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130470940>:
> 
> > @@ -199,13 +200,6 @@ enum SetResponse glSeries::set(const ParameterList &pParams)
>  
>  void glSeries::sNew()
>  {
> -  if(!_date->isValid())
> -  {
> -    QMessageBox::information( this, tr("Cannot Maintain G/L Series"),
> -                              tr("<p>You must enter a Distribution Date for this Series.") );
> -    _date->setFocus();
> -    return;
> -  }
>  
>    ParameterList params;
> This is a fine place to use GuiErrorCheck. Removing the check causes a problem - it's here to ensure that newdlg.set(params); a few lines later gets its required input parameter. Fixed This
> 
> In guiclient/glSeries.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130471088>:
> 
> > @@ -229,13 +223,6 @@ void glSeries::sNew()
>  
>  void glSeries::sEdit()
>  {
> -  if(!_date->isValid())
> -  {
> -    QMessageBox::information( this, tr("Cannot Maintain G/L Series"),
> -                              tr("<p>You must enter a Distribution Date for this Series.") );
> -    _date->setFocus();
> -    return;
> -  }
> Same here - this is an important check for valid parameter passing to a new dialog. Fixed This
> 
> 
> In guiclient/glTransaction.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130471467>:
> 
> > -      tr("<p>You must enter some Notes to describe this transaction."), _notes},
> -    { true, "", NULL }
> -  }; // error[]
> -
> -  int errIndex;
> -  for (errIndex = 0; ! error[errIndex].condition; errIndex++)
> +
> +  QList<GuiErrorCheck> errors;
> +    errors<< GuiErrorCheck(_amount->isZero(), _amount,
> +                           tr("You must enter an amount for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(! _debit->isValid(), _debit,
> +                           tr("You must select a Debit Account for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(! _credit->isValid(), _credit,
> +                           tr("You must select a Credit Account for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(!_metrics->boolean("IgnoreCompany") && _credit->companyId() != _debit->companyId(), _credit,
> +                           tr("The Accounts must belong to the same Company to Post this transaciton."))
> There's a typo in the original - transaciton - that is being propagated here. Fixed This
> 
> 
> In guiclient/glTransaction.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130471574>:
> 
> > -  }; // error[]
> -
> -  int errIndex;
> -  for (errIndex = 0; ! error[errIndex].condition; errIndex++)
> +
> +  QList<GuiErrorCheck> errors;
> +    errors<< GuiErrorCheck(_amount->isZero(), _amount,
> +                           tr("You must enter an amount for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(! _debit->isValid(), _debit,
> +                           tr("You must select a Debit Account for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(! _credit->isValid(), _credit,
> +                           tr("You must select a Credit Account for this G/L Transaction before you may Post it."))
> +          << GuiErrorCheck(!_metrics->boolean("IgnoreCompany") && _credit->companyId() != _debit->companyId(), _credit,
> +                           tr("The Accounts must belong to the same Company to Post this transaciton."))
> +          << GuiErrorCheck(_metrics->boolean("MandatoryGLEntryNotes") && _notes->toPlainText().trimmed().isEmpty(), _notes,
> +                           tr("You must enter some Notes to describe this transaction."))
> 👍 You found the intermediate style!
> 
> In guiclient/incident.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130472226>:
> 
> > -    }
> -
> -    if(_description->text().trimmed().isEmpty())
> -    {
> -      QMessageBox::critical( this, tr("Incomplete Information"),
> -        tr("You must specify a description for this incident report.") );
> -      _description->setFocus();
> +    QList<GuiErrorCheck> errors;
> +    errors<< GuiErrorCheck(_crmacct->id() == -1, _number,
> +                           tr("You must specify the Account that this incident is for."))
> +          << GuiErrorCheck(_cntct->id() <= 0 && _cntct->name().simplified().isEmpty(), _cntct,
> +                           tr("You must specify a Contact for this Incident."))
> +          << GuiErrorCheck(_description->text().trimmed().isEmpty(), _description,
> +                           tr("You must specify a description for this incident report."))
> +          << GuiErrorCheck(_status->currentIndex() == 3 && _assignedTo->username().isEmpty(), _assignedTo,
> +                           tr("You must specify an assignee when the status is assigned."))
> Nice catch on the wrong focus widget.
> 
> In guiclient/invoice.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130472441>:
> 
> > @@ -540,30 +541,20 @@ void invoice::sSave()
>      QString	msg;
>      QWidget*	widget;
>    } error[] = {
> -    { _cust->id() <= 0,
> -      tr("<p>You must enter a Customer for this Invoice before saving it."),
> -      _cust
> -    },
>      // TODO: add more error checks here?
>      { true, "", NULL }
>    };
> The whole struct definition can be removed. Fixed This
> 
> 
> In guiclient/miscCheck.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130472946>:
> 
> > -  }
> - 
> -  if ( (_expense->isChecked()) && (_expcat->id() == -1) )
> -  {
> -    QMessageBox::warning( this, tr("Cannot Create Miscellaneous Payment"),
> -                          tr("<p>You must select an Expense Category for this "
> -			     "expensed payment.") );
> -    _expcat->setFocus();
> -    return;
> -  }
> +  QList<GuiErrorCheck> errors;
> +    errors<< GuiErrorCheck(!_date->isValid(), _date,
> +                           tr("You must enter a date for this payment."))
> +          << GuiErrorCheck(_amount->isZero(), _amount,
> +                           tr("You must enter an amount for this payment."))
> +          << GuiErrorCheck((_applytocm->isChecked()) && (_cmCluster->id() == -1), _expcat,
> Hmmm... Try using _cmCluster as the focus widget. It would be a better choice here; I don't know why the old code had _expcat. Fixed This
> 
> 
> In guiclient/miscCheck.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130473307>:
> 
> >  
> -  if ( (_applytocm->isChecked()) && (_cmCluster->id() == -1) )
> -  {
> -    QMessageBox::warning( this, tr("Cannot Create Miscellaneous Payment"),
> -                          tr("<p>You must select a Credit Memo for this "
> -			     "expensed payment.") );
> -    _expcat->setFocus();
> -    return;
> -  }
> - 
> -  if (_applytocm->isChecked() && _cmCluster->isValid() && (_amt > _aropenamt))
> -  {
> -    QMessageBox::warning( this, tr("Invalid Amount"),
> -                            tr("<p>You must enter an amount less than or equal to the  "
> -	   		                   "credit memo selected.") );
> -    _amount->setLocalValue(_aropenamt);
> There's no good way to carry through this functionality to the new style. Thoughts? I had removed all of this from miscCheck and added them to the guiErrorCheck section. Did you want them to be in the new format?
> 
> 
> In guiclient/reconcileBankaccount.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130473924>:
> 
> >  
> -  if(!_datesAreOK)
> -  {
> -    QMessageBox::critical( this, tr("Dates already reconciled"),
> -                tr("The date range you have entered already has "
> -                   "reconciled dates in it. Please choose a different "
> -                   "date range.") );
> -    _startDate->setFocus();
> +  if(!_datesAreOK){
> Is there a reason to keep the if? I kept the if statement because in the previous setup if !_datesAreOK was a much longer function and at the end set datesAreOK to false so in order to keep that consistent I kept the if statement.
> 
> 
> In guiclient/standardJournal.cpp <https://github.com/xtuple/qt-client/pull/1701#discussion_r130474444>:
> 
> > @@ -137,14 +132,16 @@ void standardJournal::sSave()
>    standardSave.bindValue(":stdjrnl_name", _name->text());
>    standardSave.bindValue(":stdjrnl_id", _stdjrnlid);
>    standardSave.exec();
> -  if (standardSave.first())
> -  {
> -    QMessageBox::warning( this, tr("Cannot Save Standard Journal"),
> -                          tr("The Name you have entered for this Standard Journal already exists. "
> -                             "Please enter in a different Name for this Standard Journal."));
> -    _name->setFocus();
> +
> +  QList<GuiErrorCheck> errors;
> +  errors<< GuiErrorCheck(_name->text().length() == 0, _name,
> +                         tr("You must enter a valid Name.")),
> +  errors<< GuiErrorCheck(standardSave.first(), _name,
> +                         tr("The Name you have entered for this Standard Journal already exists. "
> +                           "Please enter in a different Name for this Standard Journal."))
> +  ;
> Nice catch with standardSave.first().
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/xtuple/qt-client/pull/1701#pullrequestreview-53350420>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHy6UkZutYdHKZw4fYbuPN73ykRPuHURks5sTk8kgaJpZM4Oo5T3>.
>

gilmoskowitz commented:
@NicholasBellucci - your comments are hard to find because the formatting from your email client got lost. The best way to respond to a code review is through the github browser interface; reply to each comment independently. It's also not necessary to add `Fixed this` comments - the code and github UI will show that after the next push :-)
